### ISSUE TITLE
Sharding strategy in 405b paper

### ISSUE THREAD

**Issue by corey-lambda** [LambdaLabsML/distributed-training-guide#38]
Quote from paper
> Our use of FSDP for Llama 3 shards optimizer states and gradients, but for model shards we do not reshard after forward computation to avoid an extra all-gather communication during backward passes.

My question is doesn't this result in having all of the model in memory by the end of forward? Not clear on how they accomplished this.

I believe this is related to `SHARD_GRAD_OP` for pytorch fsdp:

> SHARD_GRAD_OP: Gradients and optimizer states are sharded during computation, and additionally, parameters are sharded outside computation. For the parameters, this strategy unshards before the forward, does not reshard them after the forward, and only reshards them after the backward computation. The sharded optimizer states are updated locally per rank. Inside no_sync(), the parameters are not resharded after the backward computation.

**Comment by corey-lambda** [comment]
It seems we need to use device_mesh when using SHARD_GRAD_OP/HYBRID_SHARD. That was the missing piece when I tried this before
